iree/tools/android/run_module_app/src/main.cc

```diff
@@ -0,0 +1,209 @@
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <android/log.h>
+#include <android_native_app_glue.h>
+
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "iree/base/initializer.h"
+#include "iree/base/status.h"
+#include "iree/modules/hal/hal_module.h"
+#include "iree/tools/vm_util.h"
+#include "iree/vm/api.h"
+
+namespace iree {
+namespace {
+
+const char* kAppTag = "iree-run-module";
+#define LOGI(...) \
+  ((void)__android_log_print(ANDROID_LOG_INFO, kAppTag, __VA_ARGS__))
+#define LOGE(...) \
+  ((void)__android_log_print(ANDROID_LOG_ERROR, kAppTag, __VA_ARGS__))
+
+const char kModuleFileName[] = "module.vmfb";
+const char kEntryFunctionFileName[] = "entry_function.txt";
+const char kInputsFileName[] = "inputs.txt";
+const char kDriverFileName[] = "driver.txt";
+
+// A struct containing information regarding one IREE VM module invocation.
+struct IreeModuleInvocation {
+  std::string module;
+  std::string entry_function;
+  std::string inputs;
+  std::string driver;
+
+  operator bool() const {
+    return !module.empty() && !entry_function.empty() && !inputs.empty() &&
+           !driver.empty();
+  }
+};
+
+// A class for loading IREE module invocation information from Android apk asset
+// files.
+class ModuleLoader {
+ public:
+  explicit ModuleLoader(android_app* app) : app_context_(app) {}
+  ~ModuleLoader() = default;
+
+  IreeModuleInvocation LoadModuleInvocation() {
+    IreeModuleInvocation invocation = {};
+    invocation.module = ReadFileAsset(kModuleFileName);
+    invocation.entry_function = ReadFileAsset(kEntryFunctionFileName);
+    invocation.inputs = ReadFileAsset(kInputsFileName);
+    invocation.driver = ReadFileAsset(kDriverFileName);
+    return invocation;
+  }
+
+ private:
+  // Reads the given asset file and returns its contents. Returns an empty
+  // string on error.
+  std::string ReadFileAsset(const char* file_name) {
+    AAssetManager* asset_manager = app_context_->activity->assetManager;
+    AAsset* asset =
+        AAssetManager_open(asset_manager, file_name, AASSET_MODE_BUFFER);
+    if (!asset) {
+      LOGE("failed to open file '%s' in assets", kModuleFileName);
+      return {};
+    }
+
+    size_t size_in_bytes = AAsset_getLength(asset);
+    std::string contents;
+    contents.resize(size_in_bytes);
+
+    AAsset_read(asset, const_cast<char*>(contents.data()), size_in_bytes);
+    AAsset_close(asset);
+
+    return contents;
+  }
+
+  android_app* app_context_;
+};
+
+Status RunModule(const IreeModuleInvocation& invocation) {
```

Comment:
So this isn't really packing the `iree-run-module` CLI. It's just running a module, right? I would make that clear in the documentation.

Reply:
I think I made that explicit in the README. :)